### ISSUE TITLE
03. Implement `TabularCPD`, `LinearGaussianCPD`, `FunctionalCPD`

### ISSUE THREAD

**Issue by daehyun99** [pgmpy/pgmpy#3455]
It is child issue of #3452 for tracking.

- [ ] #3495
- [ ] #3496 
- [ ] #3497

### Proposal Solutions

```py
# pgmpy/parameterization/TabularCPD.py
class TabularCPD(BaseParameter):
    def __init__(
        self,
        variable_card,
        evidence_card=None,
        state_names=None,
        prior_type=None,
        equivalent_sample_size=10,
        pseudo_counts=None
    ):
    ...
    # Fitted attributes (sklearn convention)
    values_: np.ndarray             # (variable_card, prod(evidence_card))
    classes_: np.ndarray            # child state labels
    feature_names_in_: np.ndarray   # parent labels in column order (if known)
    is_fitted_: bool

    @classmethod
    def from_values(
        cls,
        variable_card,
        values,
        evidence_card=None,
        state_names=None,
        parent_order=None,
    ):
        return TabularCPD(...)

    def fit(self, X, y, sample_weight=None):
        return self

    def predict_proba(self, X):
        ...
        return CategoricalDistribution() # (len(X), variable_card)

    def sample(self, X, n_samples=1):
        dist = self.predict_proba(X)
        samples = dist.sample()
        return pd.Series(samples, ...)
```

```py
# pgmpy/parameterization/LinearGaussianCPD.py
class LinearGaussianCPD(BaseParameter):
    def __init__(self):
        ...

    beta_: np.ndarray               # length n_parents + 1; beta_[0] = intercept
    std_: float
    feature_names_in_: np.ndarray   # parent labels in beta_[1:] order (if known)

    @classmethod
    def from_values(cls, beta, std, parent_order=None):
        ...
        return LinearGaussianCPD()

    def fit(self, X, y, sample_weight=None):
        ...
        return self

    def predict_proba(self, X):
        ...
        return skpro.distributions.Normal(mu, sigma)

    def sample(self, X, n_samples=1):
        dist = self.predict_proba(X)
        samples = dist.sample()
        return pd.Series(samples, ...)

```

```py
# pgmpy/parameterization/FunctionalCPD.py
class FunctionalCPD(BaseParameter):
    def __init__(self):
        ...

    @classmethod
    def from_values(cls):
        ...
        return FunctionalCPD()

    def fit(self, X, y):
        ...
        return mu, sigma

```

### Alternative Solutions

- `TabularCPD` inherits from `ClassifierMixin`:
    - only the tag system from `ClassifierMixin` can be used.

- `log_prob` and `predict_log_proba` and (`log_pdf`, `log_pmf`)
    - scikit-learn’s `predict_log_proba()` is a classifier-specific method, and its role is different from that of `log_prob()`, `log_pdf()`, or `log_pmf()`.
    - Also, since it uses `skpro.distribution`, it seems easier to use `log_pdf` and `log_pmf`.

- ENH proposal link: https://github.com/pgmpy/enhancement_proposals/pull/12

**Comment by daehyun99** [comment]
## Suggestion 1

I don’t think `from_values()` necessarily needs to be implemented as a class method; an instance method would be sufficient.
I also think it would be better for `from_values()` to accept only attribute values generated after training, rather than hyperparameter values.

> It might be better to use a `set_values()` API, similar to `get_params()` and `set_params()`.


```py
cpd = TabularCPD(variable_card=2,state_names=[["easy", "hard"]])
cpd.from_values(values=[[0.6], [0.4]])
```

## Alternative 1

In sktime’s `FixedParams`, attribute values can be set directly through `param_dict`. These values are created as attributes when the `fit()` method is called.
(`FixedParams` is not a predictable estimator, but I thought the idea was useful, so I referenced it.)

https://github.com/sktime/sktime/blob/main/sktime/param_est/fixed.py

One advantage is that this follows the scikit-learn lifecycle, where attributes are created when `fit()` is called.
It also avoids the need to introduce a new method called `from_values()`.


```py
### example
class TabularCPD(BaseParameter):
    def __init__(
        self,
        variable_card,
        evidence_card=None,
        state_names=None,
        prior_type=None,
        equivalent_sample_size=10,
        pseudo_counts=None,
	param_dict=None # {"values": ...}
    ):

    def fi(self, X, y, sample_weight=None):
	if param_dict is not None:
		# inject param_dict value
		# self.values_ = ...

**Comment by daehyun99** [comment]
## Suggestion 2
According to the scikit-learn API, `__init__` should only accept hyperparameters that control the learning process.
The `variable_card`, `evidence_card`, and `state_names` parameters of `TabularCPD` are not hyperparameters.

Therefore, I believe only the following parameters should be exposed in `__init__`:
```python
estimator="mle"
prior_type
equivalent_sample_size
pseudo_counts
```

This API assumes that parameter estimator classes do not receive `model` and `data` information and do not orchestrate the overall training process.

Instead, training orchestration will be the responsibility of the `graph` or `model` class, as in the following API:

```python
dag.parameters.add(
    variable="X",
    cpd=TabularCPD(estimator="MLE"),
)

dag.parameters.add(
    variable="Y",
    cpd=NGBoostRegressor(),
    parent_order=["X"],
)

# orchestration
dag.fit(data)
dag.simulate(n_samples=1000)
```

> Additionally, the reason I excluded `variable_card`, `evidence_card`, and `state_names` from `TabularCPD` is as follows.

In scikit-learn classification models, the labels in `y` are stored in `self.classes_`.
The problem is that if a particular label does not appear even once in the training data, that label will not be included in `self.classes_`.

In the future implementation of `TabularCPD`, we plan to handle this case using the `*_cards` values.

I believe we need a common way to handle this across `sklearnAdapter` and `TabularCPD`.

Therefore, how about limiting the scope of this issue to cases where `*_cards` and `state_names` are inferred automatically from the data, and developing a shared handling mechanism in #3457?
